### ISSUE TITLE
PGP smartcard support

### ISSUE THREAD

**Issue by ArdaXi** [mailpile/Mailpile#926]
I just tried installing Mailpile on my local system. On starting, it immediately recognized my only GPG key and asked me for its password. Except… I don't have one, because it's on a smartcard.

In normal usage, GPG agent recognizes my key to be a smartcard and asks me for a PIN, showing me the number of sigs it's made so far. Perhaps similar functionality could be built into mailpile.

Related to issue #912 I believe.

**Comment by sneak** [comment]
+1 for gpg-agent support.

**Comment by wiml** [comment]
Another GPG-capable smartcard is the [Yubikey Neo](https://www.yubico.com/products/yubikey-hardware/yubikey-neo/); as far as I know it uses the same applet as the g10code card, so I think any solution should work with both.

**Comment by BjarniRunar** [comment]
Update: For various reasons, this issue will not be resolved until we move to GnuPG 2.1 (issue #1133).

**Comment by BjarniRunar** [comment]
Another update: This might work with the latest code after-all, in localhost configurations where you have a working agent. On the Mailpile CLI you will need to run:

mailpile> set prefs.gpg_use_agent = true

and possibly also:

mailpile> set sys.gpg_binary = /path/to/your/preferred/gpg

Of course, for remote use over the web, you're totally out of luck until we integrate GnuPG 2.1 and our own specialized agent behaviours. But smart-card users may not want remote access anyway. I'd be very interested to hear from folks whether things work or not with the above config changes.

**Comment by bnvk** [comment]
@BjarniRunar most interesting. Is the `sys.gpg_binary` basically the same thing as what Enigmail mail calls "GnuPG was found in..." as can seen in this screenshot from Qubes OS

![GnuPG path Enigmail](https://www.qubes-os.org/attachment/wiki/UserDoc/SplitGpg/tb-enigmail-split-gpg-settings-2.png)

**Comment by BjarniRunar** [comment]
@bnvk Uh, maybe? Seems likely from the labelling, but I don't know what qubes is or why gpg needs a wrapper in your case. :-)

**Comment by Kangou** [comment]
```
mailpile> set prefs.gpg_use_agent = true
Traceback (most recent call last):
  File "/Volumes/KEY/Mailpile.app/Contents/Resources/Mailpile/mailpile/commands.py", line 564, in _run_sync
    return self._finishing(command(self, *args, **kwargs))
  File "/Volumes/KEY/Mailpile.app/Contents/Resources/Mailpile/mailpile/commands.py", line 562, in command
    return self.command(*args, **kwargs)
  File "/Volumes/KEY/Mailpile.app/Contents/Resources/Mailpile/mailpile/commands.py", line 1835, in command
    cfg[var] = value
  File "/Volumes/KEY/Mailpile.app/Contents/Resources/Mailpile/mailpile/config.py", line 828, in __setitem__
    checker = self.get_rule(key)[self.RULE_CHECKER]
  File "/Volumes/KEY/Mailpile.app/Contents/Resources/Mailpile/mailpile/config.py", line 733, in get_rule
    ) % (self._name, key))
InvalidKeyError: Invalid key for config/prefs: gpg_use_agent

set error: Failed: set prefs.gpg_use_agent = true
Elapsed: 0.001s (set: Failed: set prefs.gpg_use_agent = true)

Failed: Failed: set prefs.gpg_use_agent = true

mailpile> 
```
With the last release for OS X. Same for the sys.gpg_binary.

**Comment by BjarniRunar** [comment]
@Kangou If you're not pulling from github, then you probably haven't got recent enough code for this to work. Beta III will have it... soon soon soon :)

**Comment by vlcinsky** [comment]
Few words about significance of smart card support:

Importance of private key:
- MailPile cares about privacy and security, thus using GPG encryption and signing
- private key used for signing and decryption is very sensitive material as it is a key to all the content recieved and to the ability to sign with my own key. In a way, MailPile makes that importance even bigger (by supporting it's daily use)

Researching options for protecting the key, I found smart cards to be the best option (we are using YubiKey, but smart cards in general serves the same style). Storing the private key on disk gives you an option of password protection, but the file can be possibly stolen and someone can use e.g. brute force attack to crack it.
Smart cards give excellent protection, because:
- private key is stored on the card/key
- one cannot extract the key from the card, the card only support operations (sign, decrypt...) and never reveals the private key itself
- should the card/key be stolen, the key is protected by a PIN with limited number of retries (then the content shall become unusable).
- some solutions (e.g. properly configured YubiKey) may even require user to touch a button on the key to perform some actions.

For those reasons I consider support of smart cards obligatory requirement for completed solution. On the other hand I understand, completing a solution like MailPile is complex task and it takes time.

**Comment by BjarniRunar** [comment]
@vlcinsky Yes, this will take time.

However, I suspect that keys on smart-cards will always be too cumbersome for the kind of use Mailpile has in mind. If you have to type in a pin or perform some action every single time Mailpile composes or reads an e-mail, that will quickly become such an annoyance that people will just turn the security off. 

Similarly, many smart-card based solutions completely rule out any remote access to your pile, again making Mailpile so inconvenient (compared to something like GMail which works no matter where you are, from all your devices) that many users just won't bother.

So to sum up: supporting smart-cards for the users that really want them is a "nice to have". Making it a core feature and encouraging everyone to use them? Probably never gonna happen.

Security isn't the only concern. People need to actually be willing to use the app!

**Comment by sneak** [comment]
The `gpg-agent` daemon that is responsible for talking to OpenPGP-compatible cards (dunno about pkcs11)  caches such PINs for a configurable duration so that one doesn't have to type it per-message. I use them all the time.

Yubikey 4s are excellent for this, and are $50.  I urge you to reconsider not using them as first-class options.

**Comment by danyill** [comment]
I have a Yubikey Neo and have configured it  following the instructions on the Yubikey website. I don't mind putting in a pin and this is often configurable on smart cards in any case. I use the Neo for 2FA on websites, git signing and would ideally like to at least sign my emails.

Following the advice from @BjarniRunar I set in the mailpile cli:

`set prefs.gpg_use_agent = true`

When I did this I received the normal pin request for the Yubikey from opengpg. The message was sent and was encrypted.

Unfortunately Mailpile doesn't seem willing to decrypt the message I sent to myself (and it only appears in the Sent items and not in my Inbox). But the good news is (I think) that smartcards or at least Yubikeys can work just fine!

Loving all the good work on this project.

**Comment by danyill** [comment]
On that perhaps we could confirm why Mailpile will not decrypt a message I sent to myself.

I see an attachment like this:

![image](https://user-images.githubusercontent.com/674783/29583858-ed72545a-87d5-11e7-9f20-26ef87ecdf4b.png)

I speculate that this could be because Mailpile doesn't have the private key cached and doesn't ask when viewing an email?

**Comment by vlcinsky** [comment]
Thanks @BjarniRunar , that sounds great.

@danyill consider filing extra issue if the problem with decrypting message sent to self persist (I wonder, if it relates to smartcard or it is general issue).

**Comment by BjarniRunar** [comment]
Yes, sorry I was a little abrupt closing - bugs need their own issues where we can focus on one thing at a time. Thanks for the reminder and constructive tips, @vlcinsky.
